### ISSUE TITLE
References through template instantiations are not recognized

### ISSUE THREAD

**Issue by kadircet** [llvm/llvm-project#61927]
foo.h:
```cpp
#pragma once
struct Bar;
struct Foo { Bar getBar(); };
```
bar.h:
```cpp
#pragma once
struct Bar { void bar(); };
```
use.cc:
```
#include "bar.h"
#include "foo.h"
template <typename T> void usesBar(T t) { t.getBar().bar(); }
void foo() { usesBar(Foo{}); }
```

in use.cc, we've got a false negative for the use of `Bar`.

Currently the only workaround is for people to add `// IWYU pragma: keep` to such includes.

This is a rare enough use case, but we can consider analyzing instantiated bodies when both the instantiation point and the templated body is within the same file. (we should already be visiting template parameters even when template body is in a different file, but that's not enough).

**Comment by tbaederr** [comment]
> in use.cc, we've got a false negative for the use of Bar.

False negative for what? What problem is this issue about?

**Comment by kadircet** [comment]
sorry forgot to add the label, this is about clang-include-cleaner tool from clang-tools-extra
